pktgen_s/server/pktgen_server.c

```diff
@@ -0,0 +1,331 @@
+        if ((sock = socket(p->ai_family, p->ai_socktype, p->ai_protocol)) < 0) {
+            continue;
+        }
+
+        if (connect(sock, p->ai_addr, p->ai_addrlen) == -1) {
+            close(sock);
+            continue;
+        }
+		break;
+    }
+
+    if (p == NULL) {
+        return -2;
+    }
+
+    freeaddrinfo(res); 
+
+    return sock;
+
+}
+
+int
+read_n_bytes(int sock, int n, char *buf) 
+{	
+	int bytes_read = 0;
+	while (bytes_read < n) {
+		int res = recv(sock, buf + bytes_read, n - bytes_read, 0);
+		if (bytes_read < 1) {
+			break;
+		}
+		bytes_read += res;
+	}
+	return bytes_read;
+}
+
+int
+request_handler(int fd_client, char *request)
+{
+
+	int32_t req_len;
+	char len_buf[4];
+
+	// read 4 bytes to get the length of the job
+	if (read_n_bytes(fd_client, 4, len_buf) < 0) {
+		syslog(LOG_ERR, "Failed to read length of status.");
+	}
+	
+	// convert to host
+	req_len = ntohl(*((int32_t*) &len_buf));
+
+	// read req_len bytes from the socket
+	if (read_n_bytes(fd_client, req_len, request) < 0) {
+		syslog(LOG_ERR, "Failed to read status.");
+	}
+
+	// return len of Job
+	return req_len;
+}
+
+int
+send_status(int status)
+{
+	int sock = connect_socket(SCHEDULER_IP, SCHEDULER_PORT);
+	unsigned len;
+	int32_t packed_len;
+	void *buf;
+
+	if (sock < 0) {
+		syslog(LOG_ERR, "Failed to connect to the scheduler to send status.");
+		close(sock);
+		return -1;
+	}
+
+	// set status
+	Status s = STATUS__INIT;
+	s.has_type = 1;
+	s.type = status;
+
+	// get length of serialized data
+	len = status__get_packed_size(&s);
+	buf = malloc(len+4);
+	packed_len = htonl((int32_t) len);
+
+	// set the first 4 bytes to be the length of data
+	// then pack the status into the buf
+	memcpy(buf, &packed_len, 4);
+	status__pack(&s, buf+4);
+	
+	// send the buf to the socket
+	if (send(sock, buf, len+4, 0) < 0) {
+		syslog(LOG_ERR, "Failed to send status to the scheduler.");
+		close(sock);
+		return -1;
+	}
+	// finish off
+	close(sock);
+	return 0;
+}
+
+int
+response_handler(int fd, char *request, int request_bytes)
+{
+	Job *j = job__unpack(NULL, request_bytes, request);
+
+	if (j == NULL) {
+		syslog(LOG_ERR, "Failed to unpack job.");
+		return -1;
+	}
+
+	// just print the job data for now
+	syslog(LOG_ERR, "m: %d t: %d w: %d n: %d s_min: %d "
+		            "s_max: %d r: %d l: %d.",
+		            j->m,
+		            j->t,
+		            j->w,
+		            j->n,
+		            j->s_min,
+		            j->s_max,
+		            j->r,
+		            j->l);
+
+	// sleep for 10 seconds like its actually doing something
+	sleep(10);
+
+	// unpack job
+	job__free_unpacked(j, NULL);
+	
+	// send success status regardless for now
+	return send_status(STATUS__TYPE__SUCCESS);
+}
+
+void
+sigchild_handler(int sig) 
+{
+	int tmp_errno = errno;
+	while (waitpid((pid_t)(-1), 0, WNOHANG) > 0) {}
+	errno = tmp_errno;
+}
+
+int
+main(void) 
+{
+	// setup daemon
+	setup_daemon();
+
+	syslog(LOG_NOTICE, "Pktgen Server started.");
```

Comment:
I think it would be better to use our own log file here. It would be easier to follow without having to dig through all the other stuff in syslog